### PR TITLE
Add Viterbi algorithm

### PR DESCRIPTION
**Pull request by carlos3dx**
### Describe your change:

Added Viterbi Algorithm
https://en.wikipedia.org/wiki/Viterbi_algorithm

Despite being an algorithm on Python in the article, worked my way from the mathematical definition.

To evaluate the output used the example case from the article

Fixes: #7465

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/viterbi.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    )
+
+    # The final observation
+    last_state = _extract_final_state(observations_space, states_space, probabilities)
+
+    # Process pointers backwards
+    return _extract_best_path(observations_space, last_state, pointers)
+
+
+def _validation(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:196]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        transition_probabilities,
+        emission_probabilities,
+    )
+    _validate_lists(observations_space, states_space)
+    _validate_dicts(
+        initial_probabilities, transition_probabilities, emission_probabilities
+    )
+
+
+def _validate_not_empty(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:235]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            states_space,
+            initial_probabilities,
+            transition_probabilities,
+            emission_probabilities,
+        ]
+    ):
+        raise ValueError("There's an empty parameter")
+
+
+def _validate_lists(observations_space: Any, states_space: Any) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:267]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    _validate_list(observations_space, "observations_space")
+    _validate_list(states_space, "states_space")
+
+
+def _validate_list(_object: Any, var_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:285]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    if not isinstance(_object, list):
+        raise ValueError(f"{var_name} must be a list")
+    else:
+        for x in _object:
+            if not isinstance(x, str):
+                raise ValueError(f"{var_name} must be a list of strings")
+
+
+def _validate_dicts(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:307]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    initial_probabilities: Any,
+    transition_probabilities: Any,
+    emission_probabilities: Any,
+) -> None:
+    _validate_dict(initial_probabilities, "initial_probabilities", float)
+    _validate_nested_dict(transition_probabilities, "transition_probabilities")
+    _validate_nested_dict(emission_probabilities, "emission_probabilities")
+
+
+def _validate_nested_dict(_object: Any, var_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:337]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    )
+
+    # The final observation
+    last_state = _extract_final_state(observations_space, states_space, probabilities)
+
+    # Process pointers backwards
+    return _extract_best_path(observations_space, last_state, pointers)
+
+
+def _validation(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:196]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        transition_probabilities,
+        emission_probabilities,
+    )
+    _validate_lists(observations_space, states_space)
+    _validate_dicts(
+        initial_probabilities, transition_probabilities, emission_probabilities
+    )
+
+
+def _validate_not_empty(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:235]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+            states_space,
+            initial_probabilities,
+            transition_probabilities,
+            emission_probabilities,
+        ]
+    ):
+        raise ValueError("There's an empty parameter")
+
+
+def _validate_lists(observations_space: Any, states_space: Any) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:267]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+    _validate_list(observations_space, "observations_space")
+    _validate_list(states_space, "states_space")
+
+
+def _validate_list(_object: Any, var_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:285]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    if not isinstance(_object, list):
+        raise ValueError(f"{var_name} must be a list")
+    else:
+        for x in _object:
+            if not isinstance(x, str):
+                raise ValueError(f"{var_name} must be a list of strings")
+
+
+def _validate_dicts(
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:307]: Added doctests, don't know why this comment doesn't show as outdated

File: dynamic_programming/viterbi.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    initial_probabilities: Any,
+    transition_probabilities: Any,
+    emission_probabilities: Any,
+) -> None:
+    _validate_dict(initial_probabilities, "initial_probabilities", float)
+    _validate_nested_dict(transition_probabilities, "transition_probabilities")
+    _validate_nested_dict(emission_probabilities, "emission_probabilities")
+
+
+def _validate_nested_dict(_object: Any, var_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (carlos3dx) [reply dynamic_programming/viterbi.py:337]: Added doctests, don't know why this comment doesn't show as outdated

**Comment by carlos3dx** [comment]
learned the lesson, don't split the code into smaller functions

**Review by cclauss — APPROVED** [review]
RSLGTM

**Comment by cclauss** [comment]
Please do not open new pull requests in this repo until November 1st. 2022.

**Pull request merged**